core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java

```diff
@@ -0,0 +1,251 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class IndexShardOperationsLockTests extends ESTestCase {
+
+    private static ThreadPool threadPool;
+
+    private IndexShardOperationsLock block;
+
+    @BeforeClass
+    public static void setupThreadPool() {
+        threadPool = new TestThreadPool("IndexShardOperationsLockTests");
+    }
+
+    @AfterClass
+    public static void shutdownThreadPool() {
+        ThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
+        threadPool = null;
+    }
+
+    @Before
+    public void createIndexShardOperationsLock() {
+         block = new IndexShardOperationsLock(logger, threadPool);
+    }
+
+    public void testAllOperationsInvoked() throws InterruptedException, TimeoutException, ExecutionException {
+        IndexShardOperationsLock block = new IndexShardOperationsLock(logger, threadPool);
+        int numThreads = 10;
+
+        List<PlainActionFuture<Releasable>> futures = new ArrayList<>();
+        List<Thread> operationThreads = new ArrayList<>();
+        CountDownLatch latch = new CountDownLatch(numThreads / 2);
+        for (int i = 0; i < numThreads; i++) {
+            PlainActionFuture<Releasable> future = new PlainActionFuture<>();
+            Thread thread = new Thread() {
+                public void run() {
+                    latch.countDown();
+                    block.acquire(future, ThreadPool.Names.GENERIC, true);
+                }
+            };
+            futures.add(future);
+            operationThreads.add(thread);
+        }
+
+        AtomicBoolean successFullyBlocked = new AtomicBoolean();
+
+
+        boolean blockingThreadThrowsException = randomBoolean();
+        Thread blockingThread = new Thread() {
+            public void run() {
+                final IndexShardClosedException exception = new IndexShardClosedException(new ShardId("blubb", "id", 0));
+                try {
+                    latch.await();
+                    block.blockOperations(1, TimeUnit.MINUTES, () -> {
+                        successFullyBlocked.set(true);
+                        if (blockingThreadThrowsException) {
+                            throw exception;
+                        }
+                    });
+                } catch (Exception e) {
+                    if (e != exception) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            }
+        };
+
+        blockingThread.start();
+
+        for (Thread thread : operationThreads) {
+            thread.start();
+        }
+
+        for (PlainActionFuture<Releasable> future : futures) {
+            Releasable releasable = future.get(1, TimeUnit.MINUTES);
+            assertNotNull(releasable);
+            releasable.close();
```

Comment:
this causes a deadlock when the operations that are blocked by the block are first in the future list while the last part of the future list prevents the block from being acquired. See the gist I made.